l1-contracts/contracts/bridge/asset-tracker/AssetTrackerBase.sol

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.28;
+
+import {Ownable2StepUpgradeable} from "@openzeppelin/contracts-upgradeable-v4/access/Ownable2StepUpgradeable.sol";
+import {ReentrancyGuard} from "../../common/ReentrancyGuard.sol";
+
+import {IAssetTrackerBase, MAX_TOKEN_BALANCE} from "./IAssetTrackerBase.sol";
+import {TokenBalanceMigrationData} from "../../common/Messaging.sol";
+
+import {L2_INTEROP_CENTER_ADDR, L2_TO_L1_MESSENGER_SYSTEM_CONTRACT} from "../../common/l2-helpers/L2ContractAddresses.sol";
+import {INativeTokenVaultBase} from "../ntv/INativeTokenVaultBase.sol";
+import {Unauthorized} from "../../common/L1ContractErrors.sol";
+import {IMessageRoot} from "../../bridgehub/IMessageRoot.sol";
+import {DynamicIncrementalMerkleMemory} from "../../common/libraries/DynamicIncrementalMerkleMemory.sol";
+import {SERVICE_TRANSACTION_SENDER} from "../../common/Config.sol";
+import {AssetHandlerModifiers} from "../interfaces/AssetHandlerModifiers.sol";
+import {IBridgehubBase} from "../../bridgehub/IBridgehubBase.sol";
+import {InsufficientChainBalance} from "./AssetTrackerErrors.sol";
+import {IAssetTrackerDataEncoding} from "./IAssetTrackerDataEncoding.sol";
+
+abstract contract AssetTrackerBase is
+    IAssetTrackerBase,
+    Ownable2StepUpgradeable,
+    AssetHandlerModifiers,
+    ReentrancyGuard
+{
+    using DynamicIncrementalMerkleMemory for DynamicIncrementalMerkleMemory.Bytes32PushTree;
+
+    /// @dev Maps token balances for each chain to prevent unauthorized spending across ZK chains.
+    /// NOTE: this function may be removed in the future, don't rely on it!
+    /// @dev For token origin chains, the balance starts at type(uint256).max, and decreases as withdrawals are made from the chain.
+    /// @dev On L1 the chainBalance for non origin chains equals the total supply of the token on the chain and unfinalized withdrawals to L1.
+    /// @dev On Gateway the chainBalance for non origin chains equals the total supply of the token on the chain.
+    /// @dev On non-Gateway L2s this mapping is only used to track the balance of native tokens.
+    mapping(uint256 chainId => mapping(bytes32 assetId => uint256 balance)) public chainBalance;
+
+    /// @notice Used on the L2 instead of the settlement layer
+    /// @dev Maps the migration number for each asset on the L2.
+    /// Needs to be equal to the migration number of the chain for the token to be bridgeable.
+    mapping(uint256 chainId => mapping(bytes32 assetId => uint256 migrationNumber)) public assetMigrationNumber;
+
+    /// Note, this mapping will be deprecated after V30 upgrade is finished.
```

Comment:
I dont think it will be deprecated

Reply:
we use this mapping for registering L2 native tokens for the first time:

```
        _assignMaxChainBalanceIfNeeded(data.tokenOriginChainId, data.assetId);
```

inside `receiveMigrationOnL1`

Reply:
> /// NOTE: this function may be removed in the future, don't rely on it!

I think something like that is good yes